src/dscanner/analysis/always_curly.d

```diff
@@ -0,0 +1,168 @@
+// Distributed under the Boost Software License, Version 1.0.
+//	  (See accompanying file LICENSE_1_0.txt or copy at
+//			http://www.boost.org/LICENSE_1_0.txt)
+
+module dscanner.analysis.always_curly;
+
+import dparse.lexer;
+import dparse.ast;
+import dscanner.analysis.base;
+import dsymbol.scope_ : Scope;
+
+import std.array : back, front;
+import std.algorithm;
+import std.range;
+import std.stdio;
+
+final class AlwaysCurlyCheck : BaseAnalyzer
+{
+	mixin AnalyzerInfo!"always_curly_check";
+
+	alias visit = BaseAnalyzer.visit;
+
+	///
+	this(string fileName, const(Token)[] tokens, bool skipTests = false)
+	{
+		super(fileName, null, skipTests);
+	}
+
+	void test(L,B)(L loc, B s)
+	{
+		if(!is(s == BlockStatement))
+		{
+			if(!s.tokens.empty) {
+				AutoFix af;
+				af.insertionBefore(s.tokens.front, "{");
+				af.insertionAfter(s.tokens.back, "{");
+				addErrorMessage(loc, KEY, MESSAGE, [af]);
+			} else {
+				addErrorMessage(loc, KEY, MESSAGE);
+			}
+		}
+	}
+
+    override void visit(const(IfStatement) stmt)
+    {
+		auto s = stmt.thenStatement.statement;
+		this.test(stmt, s);
+		if(stmt.elseStatement !is null) {
+			auto e = stmt.elseStatement.statement;
+			this.test(stmt.elseStatement, e);
+		}
+	}
+
+    override void visit(const(ForStatement) stmt)
+    {
+		auto s = stmt.declarationOrStatement;
+		if(s.statement !is null) {
+			this.test(stmt, s);
+		}
+	}
+
+    override void visit(const(ForeachStatement) stmt)
+    {
+		auto s = stmt.declarationOrStatement;
+		if(s.statement !is null) {
+			this.test(stmt, s);
+		}
+	}
+
+    override void visit(const(TryStatement) stmt)
+    {
+		auto s = stmt.declarationOrStatement;
+		if(s.statement !is null) {
+			this.test(stmt, s);
+		}
+
+		if(stmt.catches !is null) {
+			foreach(const(Catch) ct; stmt.catches.catches) {
+				this.test(ct, ct.declarationOrStatement);
+			}
+			if(stmt.catches.lastCatch !is null) {
+				auto sncnd = stmt.catches.lastCatch.statementNoCaseNoDefault;
+				if(sncnd !is null) {
+					this.test(stmt.catches.lastCatch, sncnd);
+				}
+			}
+		}
+	}
+
+    override void visit(const(WhileStatement) stmt)
+    {
+		auto s = stmt.declarationOrStatement;
+		if(s.statement !is null) {
+			this.test(stmt, s);
+		}
+	}
+
+    override void visit(const(DoStatement) stmt)
+    {
+		auto s = stmt.statementNoCaseNoDefault;
+		if(s !is null) {
+			this.test(stmt, s);
+		}
+	}
+
+	enum string KEY = "dscanner.style.always_curly";
+	enum string MESSAGE = "if|else|for|foreach|while|do|try|catch must be follow by a "
+		~ "BlockStatement aka. { }";
+}
+
+unittest
+{
+	import dscanner.analysis.config : StaticAnalysisConfig, Check, disabledConfig;
+	import dscanner.analysis.helpers : assertAnalyzerWarnings;
+	import std.format : format;
+	import std.stdio : stderr;
+
+	StaticAnalysisConfig sac = disabledConfig();
+	sac.always_curly_check = Check.enabled;
+
+	assertAnalyzerWarnings(q{
+		void testIf()
+		{
+			if(true) return; // [warn]: %s
```

Comment:
can you also test the range for at least one of the syntax things here to make sure the auto-fix appears where it's supposed to?

see https://github.com/dlang-community/D-Scanner/blob/a958f9ac7b1654cde0e15689fc9c60623b55e6ce/src/dscanner/analysis/auto_function.d#L224

```d
	assertAnalyzerWarnings(q{
		void testIf()
		{
			if(true) return;  /+
			^^^^^^^^ [warn]: %s +/
```
or wherever it is supposed to appear

Reply:
yeah, I was doing this inbetween my flight and several trains. I couldn't find the DRAFT button